### ISSUE TITLE
[Operation Question] How to separate truncation and matmul operations

### ISSUE THREAD

**Issue by wilyub** [secretflow/spu#672]
### Issue Type

Support

### Modules Involved

MPC protocol, SPU runtime

### Have you reproduced the bug with SPU HEAD?

Yes

### Have you searched existing issues?

Yes

### SPU Version

spu 0.9.0b1

### OS Platform and Distribution

Linux Ubuntu 22.04

### Python Version

3.11.5

### Current Behavior?

I'd like to do some latency testing for multiplication and truncation operations. Here is a short overview of how it would look:

Input x and weight y are represented using 16 bits fixed point each. Therefore, when you do a multiplication, you expect them to have 32 bits representation in fixed point. However, suppose you have a ring size of 64 bits. You could then do another 32 bits * 32 bits multiplication before overflowing your ring size of 64 bits. 

How can I do this in secretflow? In particular, I want to be able to control the fixed point representation size of each input to the matrix multiplication, and only truncate after specific multiplications (not every multiplication).

Thank you for your help!

### Standalone code to reproduce the issue

```Python
N/A
```

### Relevant log output

```Shell
N/A
```

**Comment by 6fj** [comment]
Hi @wilyub 

I don't think you could seperate matmul and truncate in Python bindings of SPU. You may have to dive into kernels of SPU runtime.

**Comment by 6fj** [comment]
I would suggest having a look at https://github.com/secretflow/spu/blob/main/REPO_LAYOUT.md.

**Comment by fionser** [comment]
@wilyub To bench the matmul only without the truncation, we can use "integer" matrix instead of the floating point matrix.

**Comment by anakinxc** [comment]
> Thanks for the update, I'll take a look at integer containers. Another thing that has been confusing me is benchmarking the time for softmax/silu/division, etc... From my reading of the literature (Puma, CipherGPT, etc..) these operations should introduce a high latency cost because they require a lot of communication. However, in my tests it seems like these operations do not add significant time to the end to end runtime of one layer in a transformer model? It seems like the largest amount of latency is attributed to the matrix multiplications instead.
> 
> I've added a screenshot of my log when I benchmarked one layer of Llama7B. The nonlinearities are implemented in jax (jax.nn.softmax and jax.nn.silu). However, as you see in the log, the vast majority of the time spent is on the mmul_aa operation (I assume this is matrix multiplication). Any idea if I've messed something up here?
> 
> Note: I'm using SEMI2K with FM64. I also did some 3PC testing with ABY3 (also FM64) and saw similar results. ![mpc_weird](https://private-user-images.githubusercontent.com/68028075/328612159-ea53ae95-58ab-4a88-bbf7-263b7f664348.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUxMjkyMDEsIm5iZiI6MTcxNTEyODkwMSwicGF0aCI6Ii82ODAyODA3NS8zMjg2MTIxNTktZWE1M2FlOTUtNThhYi00YTg4LWJiZjctMjYzYjdmNjY0MzQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA4VDAwNDE0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYzZTE3MjJjZjU4YjhjZjVhNmUyYjBiNzVjYzBmYWEzY2ZhM2YyYTI5MGM3NTAyODVmMzU4Y2ExZWY2YmYwYmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.fA3j203XZHg64v0prY0Yw-xnYDKPcMu_d9cdK53IUgU)

@Ye-D

**Comment by anakinxc** [comment]
> Please let me know if you find anything out about the above nonlinearities issue. I can also provide my code if it helps with reproducability (although like I said earlier, I just used the out of the box jnn.softmax() and jnn.silu() alongside some matrix multiplications).
> 
> One other thing that has been weird for me is the matmul benchmark. I tried two tests. One of them I did a matrix multiplication for the value weight matrix and the hidden states. (Only one matmul call). In the other test I did that matmul in addition to a matmul for hidden states and key weight matrix, and hidden states and query weight matrix. (Three matmul calls). However, the logs show the latency for the matmul to be nearly identitcal. Even weirder, the number of bytes sent/received are exactly the same. This sounds weird to me because I would expect matmul time to be tripled if we call three matmuls? I have sent a snapshot of my log and for the code. ![attn_code](https://private-user-images.githubusercontent.com/68028075/328959713-d09a44ea-66d8-462b-b9f5-2aed7b78db04.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUxODQ5MjgsIm5iZiI6MTcxNTE4NDYyOCwicGF0aCI6Ii82ODAyODA3NS8zMjg5NTk3MTMtZDA5YTQ0ZWEtNjZkOC00NjJiLWI5ZjUtMmFlZDdiNzhkYjA0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA4VDE2MTAyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNkMzA3Mjk4OGI5MDUzN2I2ZDMyMWQxYzZiMTQxYTdhYTM3NmU0ODViZDNmYjkwNDRmYmZmMTJmMzc4ZmJlYjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cQ-Pj52qdocYTs8156Yklew1v9CMSGJV7fQwPeoEFbo) ![matmul_1](https://private-user-images.githubusercontent.com/68028075/328959717-cd01f5d9-47c5-47e2-83f5-7f4018ae6ce9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUxODQ5MjgsIm5iZiI6MTcxNTE4NDYyOCwicGF0aCI6Ii82ODAyODA3NS8zMjg5NTk3MTctY2QwMWY1ZDktNDdjNS00N2UyLTgzZjUtN2Y0MDE4YWU2Y2U5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA4VDE2MTAyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJlYzI3NThmYWMxZDJjMTIyOTU4ZDAwMTdkOGJlMGVkZWJlZjEwMDg1OTc5MWNmNGYxYTlmN2E1NmI0MzhmMTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6FdFV3fHKpcHV6Qt764KKjFaJSf3c8ESdw2a7oYR5-w) ![matmul_2](https://private-user-images.githubusercontent.com/68028075/328959719-1f2eb256-7a7d-45e7-9429-c0d019fdbfb2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUxODQ5MjgsIm5iZiI6MTcxNTE4NDYyOCwicGF0aCI6Ii82ODAyODA3NS8zMjg5NTk3MTktMWYyZWIyNTYtN2E3ZC00NWU3LTk0MjktYzBkMDE5ZmRiZmIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA4VDE2MTAyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJmZTQyYTQ3NGZhNmM3NWE5NDg1NGZkMDM4NTJjYWVmNjE4NWQ3YjYzYzQ3MDk0MjJhYTc2MDc1MmY4OGRkYjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.UDA9VaeR3X7gqM_Cy65U7uVVlc2e9wCTbB5UDGDPWuk)

both `query_states` and `key_states` are unused values, so matmuls defining them are dead code and should be killed during optimizations.
